### DISCUSSION TITLE
Better Type-Safety When Using the `inject` Function

### DISCUSSION THREAD

**Discussion by PCOffline** [mobxjs/mobx#3781] [Ideas]
This is a proposal for an opt-in behaviour that introduces better type-safety when injecting stores.
The type-safety is improved in two ways:
1. When injecting stores, add types of known stores to ensure that all strings passed as arguments are valid stores
2. The function returned by `inject` takes a component as an argument. This component could have inferred prop types based on the stores injected. (e.g. if `Store1` was injected, then we know that the props have `Store1`)

I will emphasise this behaviour is opt-in, so it's not a breaking-change.
Here's a [POC](https://tsplay.dev/NVrR5N)

**Comment by kubk** [comment]
Thanks for the idea. Please check this: https://github.com/mobxjs/mobx/discussions/3741#discussioncomment-6693661

> Note: usually there is no need anymore to use Provider / inject in new code bases; most of its features are now covered by React.createContext.

**Reply by PCOffline** [reply]
Yes I'm aware, but many people are still stuck on the injection API and it's too costly to migrate, so why not add some fancy types :)

**Comment by mweststrate** [comment]
@PCOffline how would this work generically, e.g. in your PoC `MobxStores` is defined in userland code, as is `inject`. So in principle you'd still have to pass `MobXStores` as generic to `inject` anyway, which is not to different from typing the props? Would you say `inject<MobXStores>('inject')(....)` is still an significant improvement?

**Reply by PCOffline** [reply]
Hmm, I agree it seems clumsy. What I had in mind is that Mobx internally has a collection of all the stores and can produce a type from them. If that's not possible, I don't have a fancy solution for this problem.
Maybe the user can declare `MobxStores` and override `inject` to implicitly be `inject<MobxStores>`, but that's also cumbersome.
Alternatively, you can declare a type internally called `MobxStores` which, by default, is a string, and the user can override with module overrides. Again, very unintuitive, but I can't think of anything better.
